### PR TITLE
Implement basic control plane upgrade orchestration

### PR DESCRIPTION
**Pull request by ironcladlou**
This commit introduces basic support for control plane upgrades according to the
HostedCluster versioning API.

Control plane upgrades are initiated by updating the HostedCluster
`.spec.release.image` field, which is propagated to the HostedControlPlane
resource following some rules which are articulated in the test cases but not
yet codified in the API. For example, rollouts are serialized. See the test
cases for more details.

To make this work, the control plane operator has been updated to remove gating
logic which caused it to stop reconciling in reaction to fields like a "ready"
flag (which are reminiscent of state machine behavior). Now the
HostedControlPlane will always be reconciled idempotently.

The HostedControlPlane controller itself isn't architected with granular
reconciliation in mind and so there are still some spurious updates and
reconcile attempts when a no-op would have been preferred, but things should be
basically correct.

In addition, the HostedControlPlane controller is not very accurate with regards
to its status reporting, as it does no deep per-component analysis to gauge
control plane health. This weakness of reporting is reflected to some extent now
in HostedCluster (garbage in, garbage out).

This commit expands status for HostedCluster with a basic Availability condition
(see the new tests for details).

Most of the status reporting happening now is very minimal and has lots of room
for improvement to provide better decision making context for other logic. The
goal of this commit is to establish the bare minimum parts to make everything
work together and demonstrate a functional system for iteration.

### CODE REVIEW

File: api/v1alpha1/hosted_controlplane.go [hunk 1]
```diff
@@ -96,6 +96,9 @@ type HostedControlPlaneStatus struct {
 	// +kubebuilder:validation:Optional
 	Version string `json:"version,omitempty"`
```
Comment (enxebre) [inline api/v1alpha1/hosted_controlplane.go:97]: why do we need both `version` and `ReleaseImage`?

File: api/v1alpha1/hostedcluster_types.go [hunk 2]
```diff
@@ -54,6 +54,30 @@ type Release struct {
 	Image string `json:"image"`
 }
 
+type HostedClusterCondition struct {
```
Comment (enxebre) [inline api/v1alpha1/hostedcluster_types.go:57]: Is there a reason to not use core condition types?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 3]
```diff
@@ -443,31 +409,65 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+	// Compute availability and rollout progress based on the hosted control plane
+	// status
+
+	hcluster.Status.Version = computeClusterVersionStatus(hcluster, hcp)
+
+	r.Log.Info("Successfully reconciled")
+	return ctrl.Result{}, nil
+}
+
+func computeClusterVersionStatus(cluster *hyperv1.HostedCluster, controlPlane *hyperv1.HostedControlPlane) *hyperv1.ClusterVersionStatus {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:421]: nit: we should probably set a variable naming convention for better readability to differentiate between hostedCluster and capi.Cluster, e.g `hyperv1.HostedCluster = hcluster` and `capi.Cluster = cluster`

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+	"github.com/google/go-cmp/cmp"
+	configv1 "github.com/openshift/api/config/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+)
+
+func TestUpdateControlPlaneImage(t *testing.T) {
+	// TODO: the spec/status comparison of control plane is a weak check; the
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:14]: nice

File: api/v1alpha1/hostedcluster_types.go (hunk 2 above)
Comment (ironcladlou) [reply api/v1alpha1/hostedcluster_types.go:57]: I'm glad you brought this up, I'd been meaning to ask and keep forgetting... can we use a shared condition type for all our conditions on these API resources? If so, which? And are there existing libs for working with them, and if so where, etc. I have no reason to prefer distinct types, especially since we have no generics in the language. @csrwng @sjenning

File: api/v1alpha1/hosted_controlplane.go (hunk 1 above)
Comment (ironcladlou) [reply api/v1alpha1/hosted_controlplane.go:97]: This is from our original API design... my understanding is users of HostedCluster want to know the semver for the release, so it has to bubble up from here as it's resolved at runtime. It's not used for reconciliation and is just additional metadata associated with the release image (again, just my understanding). cc @csrwng @sjenning @derekwaynecarr

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 3 above)
Comment (ironcladlou) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:421]: How come we don't get to keep `cluster` for us and make CAPI be `ccluster`?! 😁

File: api/v1alpha1/hosted_controlplane.go (hunk 1 above)
Comment (ironcladlou) [reply api/v1alpha1/hosted_controlplane.go:100]: Fixed

File: api/v1alpha1/hostedcluster_types.go (hunk 2 above)
Comment (enxebre) [reply api/v1alpha1/hostedcluster_types.go]: fwiw In nodePools we are using core types and library for conditions atm.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 3 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: haha sgtm as well.

**Comment by ironcladlou** [comment]
@sjenning @enxebre @csrwng @derekwaynecarr this is ready for review, I'm down to just polishing at this point. Hopefully the test cases (both unit and e2e) are useful for evaluating whether the behavior is correctly defined.

**Comment by sjenning** [comment]
/lgtm

need to make sure we want to stick with this as the API, but this works for me
